pygeos/_bulk_creation.pyx

```diff
@@ -0,0 +1,140 @@
+# distutils: define_macros=GEOS_USE_ONLY_R_API
+
+from cpython cimport PyObject
+from cython cimport view
+
+cimport cython
+
+import numpy as np
+cimport numpy as np
+import pygeos
+
+from pygeos._geos cimport (
+    GEOSGeometry,
+    GEOSGeom_destroy_r,
+    GEOSGeom_clone_r,
+    GEOSGeom_createCollection_r,
+    GEOSGeomTypeId_r,
+    get_geos_handle,
+)
+from pygeos._pygeos_api cimport (
+    import_pygeos_c_api,
+    PyGEOS_CreateGeometry,
+    PyGEOS_GetGEOSGeometry
+)
+
+# initialize PyGEOS C API
+import_pygeos_c_api()
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def collections_1d(
+    object geometries,
+    object indices,
+    int geometry_type = 7,
+    int ndim = 2
+):
+    cdef Py_ssize_t geom_idx = 0
+    cdef Py_ssize_t coll_idx = 0
+    cdef Py_ssize_t coll_size = 0
+    cdef GEOSGeometry *geom = NULL
+    cdef GEOSGeometry *coll = NULL
+    cdef int expected_type = -1
+    cdef int expected_type_alt = -1
+    cdef int curr_type = -1
+
+    # Cast input arrays and define memoryviews for later usage
+    geometries = np.asarray(geometries, dtype=np.object)
+    if geometries.ndim != 1:
+        raise TypeError("geometries is not a one-dimensional array.")
+    cdef object[:] geometries_view = geometries
+    indices = np.asarray(indices, dtype=np.int32)
+    if indices.ndim != 1:
+        raise TypeError("indices is not a one-dimensional array.")
+    cdef int[:] indices_view = indices
+
+    cdef Py_ssize_t n_geoms = geometries_view.size
+
+    if geometries_view.size != indices_view.size:
+        raise ValueError("geometries and indices do not have equal size.")
+
+    if geometry_type == 4:  # MULTIPOINT
+        expected_type = 0
+    elif geometry_type == 5:  # MULTILINESTRING
+        expected_type = 1
+        expected_type_alt = 2
+    elif geometry_type == 6:  # MULTIPOLYGON
+        expected_type = 3
+    elif geometry_type == 7:
+        pass
+    else:
+        raise ValueError(f"Invalid geometry_type: {geometry_type}.")
+
+    if n_geoms == 0:
+        # return immediately if there are no geometries to return
+        return np.empty(shape=(0, ), dtype=np.object_)  
+
+    if np.any(indices[1:] < indices[:-1]):
+        raise ValueError("The indices array should be sorted.")  
+
+    # A temporary array for the geometries that will be given to CreateCollection.
+    # Its size equals n_geoms, which is much too large in most cases. But it will
+    # give overhead to investigate the optimal size at this point.
+    temp_geoms = np.empty(shape=(n_geoms, ), dtype=np.intp)
```

Comment:
I like the idea, but after doing some timings I decided to go for `np.bincount` instead of your suggestion.

```
In [19]: %timeit np.unique(arr, return_counts=True)[1]
1.97 ms ± 12 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [16]: def counts_2(arr):
    ...:     run_start = np.r_[1, arr[:-1] != arr[1:]]
    ...:     return np.diff(np.r_[np.nonzero(run_start)[0], len(arr)])
In [20]: %timeit counts_2(arr)
643 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [30]: %timeit np.bincount(arr)
239 µs ± 377 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

It has the added advantage of returning `0` for indices that do not occur:

```
np.bincount([1, 1, 5, 5])
Out[31]: array([0, 2, 0, 0, 0, 2])
```